pkg/webhook/mutation/pod/metadata/workload.go

```diff
@@ -46,43 +39,29 @@ func findRootOwnerOfPod(ctx context.Context, clt client.Client, pod *corev1.Pod,
 			Kind:       pod.Kind,
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name: pod.ObjectMeta.Name,
+			Name: kubeobjects.GetName(*pod),
 			// pod.ObjectMeta.Namespace is empty yet
 			Namespace:       namespace,
 			OwnerReferences: pod.ObjectMeta.OwnerReferences,
 		},
 	}
 
-	workloadInfo, err := findRootOwner(ctx, clt, podPartialMetadata)
+	workloadInfo, err := findRootOwner(ctx, clt, podPartialMetadata, podPartialMetadata) // default owner of the pod is the pod itself
 	if err != nil {
 		return nil, errors.WithStack(err)
 	}
 
 	return &workloadInfo, nil
 }
 
-func findRootOwner(ctx context.Context, clt client.Client, partialObjectMetadata *metav1.PartialObjectMetadata) (workloadInfo, error) {
-	if len(partialObjectMetadata.ObjectMeta.OwnerReferences) == 0 {
-		if partialObjectMetadata.ObjectMeta.Name == "" {
-			// pod is not created directly and does not have an owner reference set
-			return newUnknownWorkloadInfo(), nil
-		}
-
-		return newWorkloadInfo(partialObjectMetadata), nil
+func findRootOwner(ctx context.Context, clt client.Client, rootObject *metav1.PartialObjectMetadata, wellKnownOwner *metav1.PartialObjectMetadata) (workloadInfo, error) {
```

Comment:
IMO this is an over-complicated way of solving the problem. (you are trying to reuse the old implementation to solve a new problem, instead of writing a solution to our current problem)

**As far as I understand the problem:**
The `OwnerReferences` is a tree, where the root is the `owner` that started everything.

What we want is to return the `owner` in the tree that is:
1. a "controller"
2. "well-known"
3. closest to the root 

**A possible way to go about finding the desired owner:**
W need to do is just reach the root of the tree (where we have no more `OwnerReferences`), with the following conditions:
- only go "down" a level if we find a "controller" (1. condition satisfied)

When we're coming back from the root we have to make sure that we only return an `owner` if its "well-known" (2. condition satisfied)
- This will cause the `owner` closest to the root to be returned at the end. (3. condition satisfied)